kyaml/kyaml.go

```diff
@@ -0,0 +1,712 @@
+func (ky *Encoder) renderScalar(node *yaml.Node, indent int, flags flagMask, out io.Writer) error {
+	switch node.Tag {
+	case intTag, floatTag, boolTag, nullTag:
+		fmt.Fprint(out, node.Value)
+	case strTag:
+		return ky.renderString(node.Value, indent+1, flags, out)
+	case timestampTag:
+		return ky.renderString(node.Value, indent+1, flagsNone, out)
+	default:
+		return fmt.Errorf("kyaml internal error: line %d: unknown tag %q on scalar node %q", node.Line, node.Tag, node.Value)
+	}
+	return nil
+}
+
+const kyamlFoldStr = "\\\n"
+
+// renderString processes a string (either single-line or multi-line),
+// rendering it to the output.  This DOES NOT render a trailing newline.
+func (ky *Encoder) renderString(val string, indent int, flags flagMask, out io.Writer) error {
+	lazyQuote := flags&flagLazyQuote != 0
+	compact := flags&flagCompact != 0
+
+	// If no newlines, just use standard Go quoting.
+	if compact || !strings.Contains(val, "\n") {
+		if lazyQuote && !needsQuotes(val) {
+			fmt.Fprint(out, val)
+		} else {
+			fmt.Fprint(out, strconv.Quote(val))
+		}
+		return nil
+	}
+
+	// The input has at least one newline. We will use YAML's line folding to
+	// make the output more readable.
+	//
+	// The rest of this is borrowed from Go's strconv.Quote implementation.
+
+	s := val
+
+	// accumulate into a buffer
+	buf := &bytes.Buffer{}
+
+	// opening quote and fold
+	fmt.Fprint(buf, `"`, kyamlFoldStr)
+
+	// Iterating a string with invalid UTF8 returns RuneError rather than the
+	// bytes, so we iterate the string and decode the runes. This is a bit
+	// slower, but gives us a better result.
+	for width := 0; len(s) > 0; s = s[width:] {
+		r := rune(s[0])
+		width = 1
+		if r >= utf8.RuneSelf {
+			r, width = utf8.DecodeRuneInString(s)
+		}
+		if width == 1 && r == utf8.RuneError {
+			fmt.Fprint(buf, `\x`)
+			fmt.Fprintf(buf, "%02x", s[0])
+			continue
+		}
+		ky.appendEscapedRune(r, indent, buf)
+	}
+
+	// closing quote
+	afterNewline := buf.Bytes()[len(buf.Bytes())-1] == '\n'
+	if !afterNewline {
+		fmt.Fprint(buf, kyamlFoldStr)
+	}
+	ky.writeIndent(indent, buf)
+	fmt.Fprint(buf, `"`)
+
+	fmt.Fprint(out, buf.String())
+
+	return nil
+}
+
+var allowedUnquotedAnywhere = map[rune]bool{
+	'_': true,
+}
+
+var allowedUnquotedInterior = map[rune]bool{
+	'-': true,
+	'.': true,
+	'/': true,
+}
+
+func needsQuotes(s string) bool {
+	if s == "" {
+		return true
+	}
+	if yamlGuessesWrong(s) {
+		return true
+	}
+	runes := []rune(s)
+	for i, r := range runes {
+		if unicode.IsLetter(r) || unicode.IsNumber(r) || allowedUnquotedAnywhere[r] {
+			continue
+		}
+		if i > 0 && i < len(runes)-1 && allowedUnquotedInterior[r] {
+			continue
+		}
+		// it's something we don't explicitly allow
+		return true
+	}
+	return false
+}
+
+// From https://yaml.org/type/int.html and https://yaml.org/type/float.html
+var sexagesimalRE = regexp.MustCompile(`^[+-]?[1-9][0-9_]*(:[0-5]?[0-9])+(\.[0-9_]*)?$`)
+
+func yamlGuessesWrong(s string) bool {
```

Comment:
I know what you mean by this, but a doc comment might be helpful here:
```suggestion
// yamlGuessesWrong returns true if a string would likely parse as a YAML type other than string,
// in which case we will render it as a quoted / escaped string
func yamlGuessesWrong(s string) bool {
```

Reply:
Will rename and comment